man/man3/initclimate.3
.TH initclimate 3  "LPJmL programmers manual"
.SH NAME
initclimate \- initialize climate data
.SH SYNOPSIS
.nf
\fB#include "lpj.h"

Climate *initclimate(const Cell \fIgrid[]\fB,const Config *\fIconfig\fB);\fP

.fi
.SH DESCRIPTION
Function initializes climate data and opens related files specified in the LPJmL configuration.
The arguments of \fBinitclimate\fP are:
.TP
.I grid
Cell grid array allocated and initialized by \fBnewgrid()\fP.
.TP
.I config
Pointer to LPJmL configuration data. Must be read by \fBreadconfig()\fP.
.SH RETURN VALUE
Upon successful completion pointer to allocated climate data is returned and NULL in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
getclimate(3), freeclimate(3), readconfig(3), iterate(3) 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

## [5.6.25] - 2023-07-29

### Added

- New outputs added for N fluxes from timber extraction and for N pools in wood product pools
- New outputs added for `estab_storage` C and N
- Added missing N fluxes, `estab_storage`, and reservoirs to `globalflux` output
- Added new fluxes to man pages

### Fixed 

- Corrected/added units in comments in output.h
- Corrected units in `flux_sum` man page
- Corrected `N2O_DENIT` output, which now no longer also includes `N2O_NIT` values
- Corrected writing of mineral N pools to outputs, so that also mineral N in frozen soil layers is included
- Corrected writing of `VEGN` for crops, so that `bm_inc.nitrogen` is no longer included to avoid double accounting

### Removed

- Removed unecessary initialization of output variables to zero in `update_annual.c`, which is done in `initoutputdata.c`


## [5.6.24] - 2023-07-29

### Fixed

-  The number of bands is now correctly checked for the `read_residue_data` input.


## [5.6.23] - 2023-07-28

### Changed

- Instead of the soil code the index of the grid cell is written into the NetCDF file for `"grid"` output.


## [5.6.22] - 2023-07-27

### Changed

- Configuration files `lpjml_netcdf.js`, `lpjml_non.js`, `param_non.js`, and `lpjparam_non.js` updated to latest lpjml version.

### Removed

- Predefined daily outputs (`"d_lai"`-`"d_pet"`) and corresponding functions removed.

- Version numbers removed from man pages.

### Fixed

- Variable `issocket` always initialized to `FALSE` in `readfilename.c`.


## [5.6.21] - 2023-07-21

### Changed

- Flag `"nitrogen_coupled"` disabled for lpjml offline runs. Water stress is always coupled with nitrogen stress. Only if lpjml is compiled with `-DCOUPLING_WITH_FMS` flag is enabled.

- Flag `"new_trf"` renamed to `"transp_suction_fcn"`.


## [5.6.20] - 2023-07-13

### Added

- New flags `"fix_co2"` and `"fix_co2_year"` added.

### Changed

- Flag `"istimber"` renamed to `"luc_timber"`.

- Flag `"new_phenology"` renamed to `"gsi_phenology"`.

- Flag `"shuffle_climate"` renamed to `"shuffle_spinup_climate"`.

- The settings for fixed input data have been updated. For climate and N deposition input an interval can be specified from which data is taken after the specified year. In can specified whether random shuffling or cycling is used:

```java
"fix_climate" : false,                /* fix climate after specified year */
"fix_climate_year" : 1901,            /* year after climate is fixed */
"fix_climate_interval" : [1901,1930],
"fix_climate_shuffle" : true,          /* randomly shuffle climate in the interval */
"fix_deposition_with_climate" : false, /* fix N deposition same as climate */
"fix_deposition" : false,              /* fix N deposition after specified year */
"fix_deposition_year" : 1901,          /* year after deposition is fixed */
"fix_deposition_interval" : [1901,1930],
"fix_deposition_shuffle" : true,       /* randomly shuffle depositions in the interval */
"fix_landuse" : false,                 /* fix land use after specfied year */
"fix_landuse_year" : 1901,             /* year after land use is fixed */
"fix_co2" : false,                     /* fix atmospheric CO2  after specfied year */
"fix_co2_year" : 1901,                 /* year after CO2 is fixed */
```

### Removed

- Obsolete flag `"const_climate"` removed.


## [5.6.19] - 2023-07-06

### Added

- Help option `-h` added to `bin2cdf` and `clm2cdf` utility.

- Target `lpjcheck` added to `Makefile` to create only the `lpjcheck` utility.

### Changed

- The version of all man pages has been set to the LPJmL version number, date of last modification removed.

- The different versions of the utilities have been replaced by a reference to the LPJmL version.

### Removed

- All html files have been removed from the repository.

- Obsolete link `lpj` to `lpjml` removed.


## [5.6.18] - 2023-07-04

### Fixed

- Soil array `wpwp` is now correctly initialized with `soilpar->wpwp` for the `"prescribed_soilpar"` setting.


## [5.6.17] - 2023-06-30

### Fixed

- Replaced logical `or` by logical `and` in `survive.c`. This solves issue [#309](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/309).


## [5.6.16] - 2023-06-30

### Fixed

- Removed excess `pft->nind` in all harvest routines in `harvest_stand.c`.


## [5.6.15] - 2023-06-30

### Added

- Check for matching size of discharge queues read from restart file added to `initriver()` function in `initdrain.c`. If the size differs from the value calculated from the river lengths the queues will be resized and initialized to zero. Error message is additionally printed.


## [5.6.14] - 2023-06-08

### Removed

- Support for the deprecated `*.conf` configuration file format has been removed. Only JSON files are allowed for configuration files and metafiles for input.

- Man pages for the description of the `*.conf` file format removed.

### Changed

- Strings in JSON configuration files can now be longer than 256 characters.


## [5.6.13] - 2023-05-16

### Added


src/climate/initclimate.c
/**************************************************************************************/
/**                                                                                \n**/
/**               i  n  i  t  c  l  i  m  a  t  e  .  c                            \n**/
/**                                                                                \n**/
/**     C implementation of LPJmL                                                  \n**/
/**                                                                                \n**/
/**     Function initclimate allocates and initializes the climate data            \n**/
/**     pointer. The corresponding data files are opened and storage               \n**/
/**     allocated.                                                                 \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

static void initdata(Climate *climate)
{
  int i;
  climate->file_temp.isopen=FALSE;
  climate->file_prec.isopen=FALSE;
  climate->file_lwnet.isopen=FALSE;
  climate->file_swdown.isopen=FALSE;
  climate->file_cloud.isopen=FALSE;
  climate->file_wet.isopen=FALSE;
  climate->file_no3deposition.isopen=FALSE;
  climate->file_nh4deposition.isopen=FALSE;
  climate->file_wind.isopen=FALSE;
  climate->file_humid.isopen=FALSE;
  climate->file_tmin.isopen=FALSE;
  climate->file_tmax.isopen=FALSE;
  climate->file_tamp.isopen=FALSE;
  climate->file_lightning.isopen=FALSE;
  climate->file_burntarea.isopen=FALSE;
  climate->file_delta_temp.isopen=FALSE;
  climate->file_delta_prec.isopen=FALSE;
  climate->file_delta_lwnet.isopen=FALSE;
  climate->file_delta_swdown.isopen=FALSE;
#if defined IMAGE && defined COUPLED
  climate->file_temp_var.isopen=FALSE;
  climate->file_prec_var.isopen=FALSE;
#endif
  climate->co2.data=NULL;
  climate->ch4.data=NULL;
  for (i = 0; i<4; i++)
  {
    climate->data[i].tmax=NULL;
    climate->data[i].tmin=NULL;
    climate->data[i].prec=NULL;
    climate->data[i].temp=NULL;
    climate->data[i].sun=NULL;
    climate->data[i].lwnet=NULL;
    climate->data[i].swdown=NULL;
    climate->data[i].wet=NULL;
    climate->data[i].wind=NULL;
    climate->data[i].humid=NULL;
    climate->data[i].burntarea=NULL;
    climate->data[i].tamp=NULL;
    climate->data[i].no3deposition=NULL;
    climate->data[i].nh4deposition=NULL;
    climate->data[i].lightning=NULL;
  }
} /* of 'initdata' */

Climate *initclimate(const Cell grid[],   /**< LPJ grid */
                     const Config *config /**< pointer to LPJ configuration */
                    )                     /** \return allocated climate data struct or NULL on error */
{
  char *name;
  int i, ndata; 
  int lastyear;
  Climate *climate;
  climate=new(Climate);
  if(climate==NULL)
  {
    printallocerr("climate");
    return NULL;
  }
  initdata(climate);
  if(openclimate(&climate->file_temp,&config->temp_filename,"celsius",LPJ_SHORT,1,0.1,config))
  {
    if(isroot(*config))
      fprintf(stderr,"ERROR236: Cannot open temp data file.\n");
    freeclimate(climate,isroot(*config));
    return NULL;
  }
  climate->firstyear=climate->file_temp.firstyear;
  ndata = (config->delta_year>1) ? 3 : 1;
  if (config->isanomaly)
  {
    ndata++;
    if (openclimate(&climate->file_delta_temp, &config->delta_temp_filename, "celsius", LPJ_SHORT,config->delta_year,0.1,config))
    {
      closeclimatefile(&climate->file_temp, isroot(*config));
      free(climate);
      return NULL;
    }
    if (openclimate(&climate->file_delta_prec, &config->delta_prec_filename, "kg/m2/day" /* "mm" */, LPJ_SHORT,config->delta_year,1.0,config))
    {
      closeclimatefile(&climate->file_temp, isroot(*config));
      closeclimatefile(&climate->file_delta_temp, isroot(*config));
      free(climate);
      return NULL;
    }
    if (openclimate(&climate->file_delta_lwnet, &config->delta_lwnet_filename, "W/m2", LPJ_SHORT,config->delta_year, 0.1,config))
    {
      closeclimatefile(&climate->file_temp, isroot(*config));
      closeclimatefile(&climate->file_prec, isroot(*config));
      free(climate);
      return NULL;
    }
    if (openclimate(&climate->file_delta_swdown, &config->delta_swdown_filename, "W/m2", LPJ_SHORT,config->delta_year,0.1,config))
    {
      closeclimatefile(&climate->file_temp, isroot(*config));
      closeclimatefile(&climate->file_prec, isroot(*config));
      closeclimatefile(&climate->file_lwnet, isroot(*config));
      free(climate);
      return NULL;
    }
  }
  if(openclimate(&climate->file_prec,&config->prec_filename,"kg/m2/day" /* "mm" */,LPJ_SHORT,1,1.0,config))
  {
    if(isroot(*config))
      fprintf(stderr,"ERROR236: Cannot open prec data file.\n");
    freeclimate(climate,isroot(*config));
    return NULL;
  }
  if(climate->firstyear<climate->file_prec.firstyear)
    climate->firstyear=climate->file_prec.firstyear;
  if(config->with_radiation)
  {
    if(config->with_radiation==RADIATION || config->with_radiation==RADIATION_LWDOWN)
    {
      if(openclimate(&climate->file_lwnet,&config->lwnet_filename,"W/m2",LPJ_SHORT,1,0.1,config))
      {
        if(isroot(*config))
          fprintf(stderr,"ERROR236: Cannot open %s data file.\n",(config->with_radiation==RADIATION) ? "lwnet" : "lwdown");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
      if(climate->firstyear<climate->file_lwnet.firstyear)
        climate->firstyear=climate->file_lwnet.firstyear;
    }
    if(openclimate(&climate->file_swdown,&config->swdown_filename,"W/m2",LPJ_SHORT,1,0.1,config))
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR236: Cannot open swdown data file.\n");
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if(climate->firstyear<climate->file_swdown.firstyear)
      climate->firstyear=climate->file_swdown.firstyear;
  }
  else
  {
    if(openclimate(&climate->file_cloud,&config->cloud_filename,"%",LPJ_SHORT,1,1.0,config))
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR236: Cannot open cloud data file.\n");
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if(climate->firstyear<climate->file_cloud.firstyear)
      climate->firstyear=climate->file_cloud.firstyear;
  }
  if(config->wet_filename.name!=NULL)
  {
    if(openclimate(&climate->file_wet,&config->wet_filename,NULL,LPJ_SHORT,1,1.0,config))
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR236: Cannot open wet data file.\n");
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if(climate->firstyear<climate->file_wet.firstyear)
      climate->firstyear=climate->file_wet.firstyear;
    climate->file_wet.ready=FALSE;
  }

  if(config->with_nitrogen && config->with_nitrogen!=UNLIM_NITROGEN && !config->no_ndeposition)
  {
    if(openclimate(&climate->file_no3deposition,&config->no3deposition_filename,"g/m2/day",LPJ_FLOAT,1,1.0,config))
    {
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if(climate->firstyear<climate->file_no3deposition.firstyear)
      climate->firstyear=climate->file_no3deposition.firstyear;
    if(openclimate(&climate->file_nh4deposition,&config->nh4deposition_filename,"g/m2/day",LPJ_FLOAT,1,1.0,config))
    {
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if(isroot(*config))
    {
      lastyear=(config->fix_deposition) ? max(config->fix_deposition_year,config->fix_deposition_interval[1]) : config->lastyear;
      if(climate->file_no3deposition.firstyear+climate->file_no3deposition.nyear-1<lastyear)
        fprintf(stderr,"WARNING024: Last year in '%s'=%d is less than last simulation year %d, data from last year used.\n",
                config->no3deposition_filename.name,climate->file_no3deposition.firstyear+climate->file_nh4deposition.nyear-1,lastyear);
      if(climate->file_nh4deposition.firstyear+climate->file_nh4deposition.nyear-1<lastyear)
        fprintf(stderr,"WARNING024: Last year in '%s'=%d is less than last simulation year %d, data from last year used.\n",
                config->nh4deposition_filename.name,climate->file_nh4deposition.firstyear+climate->file_nh4deposition.nyear-1,lastyear);
    }
  }

  if(config->fire==SPITFIRE || config->fire==SPITFIRE_TMAX || config->with_nitrogen)
  {
    if(openclimate(&climate->file_wind,&config->wind_filename,"m/s",LPJ_SHORT,1,0.001,config))
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR236: Cannot open wind data file.\n");
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if(climate->firstyear<climate->file_wind.firstyear)
      climate->firstyear=climate->file_wind.firstyear;
  }

  if(config->fire==SPITFIRE || config->fire==SPITFIRE_TMAX)
  {
    if(config->fdi==WVPD_INDEX)
    {
      if(openclimate(&climate->file_humid,&config->humid_filename,NULL,LPJ_SHORT,1,1.0,config))
      {
        if(isroot(*config))
          fprintf(stderr,"ERROR236: Cannot open humid data file.\n");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
    }
  }

  if(config->cropsheatfrost || config->fire==SPITFIRE_TMAX)
  {
    if(openclimate(&climate->file_tmin,&config->tmin_filename,"celsius",LPJ_SHORT,1,0.1,config))
    {
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if(openclimate(&climate->file_tmax,&config->tmax_filename,"celsius",LPJ_SHORT,1,0.1,config))
    {
      freeclimate(climate,isroot(*config));
      return NULL;
    }
  }

  if(config->fire==SPITFIRE)
  {
    if(openclimate(&climate->file_tamp,&config->tamp_filename,NULL,LPJ_SHORT,1,0.1,config))
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR236: Cannot open tamp data file.\n");
      freeclimate(climate,isroot(*config));
      return NULL;
    }
  }


  if(config->fire==SPITFIRE || config->fire==SPITFIRE_TMAX)
  {
    if(openclimate(&climate->file_lightning,&config->lightning_filename,"1/day/hectare",LPJ_INT,1,1.0e-7,config))
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR236: Cannot open lightning data file.\n");
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if (config->prescribe_burntarea)
    {
      if(openclimate(&climate->file_burntarea,&config->burntarea_filename,(config->burntarea_filename.fmt==CDF) ?  (char *)NULL : (char *)NULL,LPJ_SHORT,1,100.0,config))
      {
        if(isroot(*config))
          fprintf(stderr,"ERROR236: Cannot open burntarea data file.\n");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
    }
  }

#if defined IMAGE && defined COUPLED
  if(config->sim_id==LPJML_IMAGE)
  {
    if(openclimate(&climate->file_temp_var,&config->temp_var_filename,NULL,LPJ_SHORT,1,0.1,config))
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR236: Cannot open tempvar data file.\n");
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if(openclimate(&climate->file_prec_var,&config->prec_var_filename,NULL,LPJ_SHORT,1,0.01,config))
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR236: Cannot open precvar data file.\n");
      freeclimate(climate,isroot(*config));
      return NULL;
    }
  }
#endif
  if(!config->isanomaly && (config->sim_id == LPJML || config->sim_id == LPJ) && climate->firstyear>config->firstyear)
  {
    if(isroot(*config))
      fprintf(stderr,"ERROR200: Climate data starts at %d, later than first simulation year %d.\n",
              climate->firstyear,config->firstyear);
    freeclimate(climate,isroot(*config));
    return NULL;
  }
  if(readtracegas(&climate->co2,&config->co2_filename,config))
  {
    freeclimate(climate,isroot(*config));
    return NULL;
  }
  if(isroot(*config) && climate->co2.firstyear>climate->file_temp.firstyear)
      fprintf(stderr,"WARNING001: First year in '%s'=%d greater than climate->file_temp.firstyear=%d.\n"
              "            Preindustrial value=%g for CO2 is used.\n",
              config->co2_filename.name,climate->co2.firstyear,climate->file_temp.firstyear,param.co2_p);
#ifdef DEBUG7
  printf("climate->file_temp.firstyear: %d  co2-year: %d  value: %f\n",
         climate->file_temp.firstyear, climate->co2.firstyear,climate->co2.data[0]);
#endif
  if (!config->with_dynamic_ch4)
  {
    if (readtracegas(&climate->ch4, &config->ch4_filename,config))
    {
      freeclimate(climate,isroot(*config));
      return NULL;
    }
    if (isroot(*config) && climate->ch4.firstyear>climate->file_temp.firstyear)
      fprintf(stderr, "WARNING001: First year in '%s'=%d greater than climate->file_temp.firstyear=%d.\n"
        "            value=%g for CH4 is used.\n",
        config->ch4_filename.name, climate->ch4.firstyear, climate->file_temp.firstyear,climate->ch4.data[0]);
  }
  if(config->prec_filename.fmt!=FMS)
  {
    for (i = 0; i<ndata; i++)
      if((climate->data[i].prec=newvec(Real,climate->file_prec.n))==NULL)
      {
        printallocerr("prec");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
  }
  if(config->temp_filename.fmt!=FMS)
  {
    for (i = 0; i<ndata; i++)
      if((climate->data[i].temp=newvec(Real,climate->file_temp.n))==NULL)
      {
        printallocerr("temp");
        freeclimate(climate,isroot(*config));
        return NULL;
    }
  }
  if(config->with_nitrogen && config->with_nitrogen!=UNLIM_NITROGEN && !config->no_ndeposition)
  {
    for (i = 0; i<ndata; i++)
      if((climate->data[i].no3deposition=newvec(Real,climate->file_no3deposition.n))==NULL)
      {
        printallocerr("no3deposition");
        freeclimate(climate,isroot(*config));
        free(climate);
        return NULL;
      }
    for (i = 0; i<ndata; i++)
      if((climate->data[i].nh4deposition=newvec(Real,climate->file_nh4deposition.n))==NULL)
      {
        printallocerr("nh4deposition");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
  }
  if(config->with_nitrogen || config->fire==SPITFIRE || config->fire==SPITFIRE_TMAX)
  {
    if(config->wind_filename.fmt!=FMS)
    {
      for (i = 0; i<ndata; i++)
        if((climate->data[i].wind=newvec(Real,climate->file_wind.n))==NULL)
        {
          printallocerr("wind");
          freeclimate(climate,isroot(*config));
          return NULL;
        }
    }
  }

  if(config->cropsheatfrost || config->fire==SPITFIRE_TMAX)
  {
    for (i = 0; i<ndata; i++)
    {
      if((climate->data[i].tmax=newvec(Real,climate->file_tmax.n))==NULL)
      {
        printallocerr("tmax");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
      if((climate->data[i].tmin=newvec(Real,climate->file_tmin.n))==NULL)
      {
        printallocerr("tmin");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
    }
  }

  if(config->fire==SPITFIRE)
  {
    if(config->tamp_filename.fmt!=FMS)
    {
      for (i = 0; i<ndata; i++)
        if((climate->data[i].tamp=newvec(Real,climate->file_tamp.n))==NULL)
        {
          printallocerr("tamp");
          freeclimate(climate,isroot(*config));
          free(climate);
          return NULL;
      }
    }
  }
  if(config->fire==SPITFIRE || config->fire==SPITFIRE_TMAX)
  {
    if(config->fdi==WVPD_INDEX)
    {
      for (i = 0; i<ndata; i++)
        if((climate->data[i].humid=newvec(Real,climate->file_humid.n))==NULL)
        {
          printallocerr("humid");
          freeclimate(climate,isroot(*config));
          return NULL;
        }
    }
    for (i = 0; i<ndata; i++)
      if((climate->data[i].lightning=newvec(Real,climate->file_lightning.n))==NULL)
      {
        printallocerr("lightning");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
    for (i = 0; i<ndata; i++)
      if(readclimate(&climate->file_lightning,climate->data[i].lightning,0.0,climate->file_lightning.scalar,grid,climate->file_lightning.firstyear,1,config))
      {
        if(isroot(*config))
          fprintf(stderr,"ERROR192: Cannot read lightning.\n");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
  } /* of if(config->fire==SPITFIRE || config->fire==SPITFIRE_TMAX) */
  if(config->with_radiation)
  {
    if(config->with_radiation==RADIATION || config->with_radiation==RADIATION_LWDOWN)
    {
      if(config->lwnet_filename.fmt!=FMS)
      {
        for (i = 0; i<ndata; i++)
          if((climate->data[i].lwnet=newvec(Real,climate->file_lwnet.n))==NULL)
          {
            printallocerr("lwnet");
            freeclimate(climate,isroot(*config));
            return NULL;
          }
      }
    }
    if(config->swdown_filename.fmt!=FMS)
    {
      for (i = 0; i < ndata; i++)
        if((climate->data[i].swdown=newvec(Real,climate->file_swdown.n))==NULL)
        {
          printallocerr("swdown");
          freeclimate(climate,isroot(*config));
          return NULL;
        }
    }
  }
  else
  {
    if(config->cloud_filename.fmt!=FMS)
    {
      for (i = 0; i < ndata; i++)
        if((climate->data[i].sun=newvec(Real,climate->file_cloud.n))==NULL)
        {
          printallocerr("cloud");
          freeclimate(climate,isroot(*config));
          return NULL;
        }
    }
  }
  if(config->wet_filename.name!=NULL && config->wet_filename.fmt!=FMS)
  {
    if(config->wet_filename.fmt!=FMS)
    {
      for (i = 0; i < ndata; i++)
        if((climate->data[i].wet=newvec(Real,climate->file_wet.n))==NULL)
        {
          printallocerr("wet");
          freeclimate(climate,isroot(*config));
          return NULL;
        }
    }
  }
  if((config->fire==SPITFIRE || config->fire==SPITFIRE_TMAX) && config->prescribe_burntarea)
  {
    for (i = 0; i < ndata; i++)
      if((climate->data[i].burntarea=newvec(Real,climate->file_burntarea.n))==NULL)
      {
        printallocerr("burntarea");
        freeclimate(climate,isroot(*config));
        return NULL;
      }
  }

  return climate;
} /* of 'initclimate' */
